src/core/exec-invoke.c

```diff
@@ -4031,6 +4025,39 @@ static int send_handoff_timestamp(
         return 1;
 }
 
+static void prepare_terminal(
+                const ExecContext *context,
+                ExecParameters *p) {
+
+        _cleanup_close_ int lock_fd = -EBADF;
+
+        /* This is the "constructive" reset, i.e. is about preparing things for our invocation rather than
+         * cleaning up things from older invocations. */
+
+        assert(context);
+        assert(p);
+
+        /* We only try to reset things if we there's the chance our stdout points to a TTY */
+        if (!(is_terminal_output(context->std_output) ||
+              (context->std_output == EXEC_OUTPUT_INHERIT && is_terminal_input(context->std_input)) ||
+              context->std_output == EXEC_OUTPUT_NAMED_FD ||
+              p->stdout_fd >= 0))
```

Comment:
I still don't understand why this can't live in setup_output, but it works for me I guess.

Reply:
It feels weird to me to stuff this in setup_output(), given that the terminal dimension stuff really needs both input and output. Makes more sense to me to try it after we set both up.